#1. Defining Problem Statement and Analysing basic metrics (10 Points)

##Download the dataset and observe a subset of the data

**Importing Required Libraries**

In [10]:
pip install numpy


SyntaxError: invalid syntax (3098743027.py, line 1)

In [6]:
import numpy as np
import pandas as pd
import seaborn as sns

ModuleNotFoundError: No module named 'numpy'

**Download the Netflix DataSet**

In [5]:
# !gdown https://d2beiqkhq929f0.cloudfront.net/public_assets/assets/000/000/940/original/netflix.csv -O netflix_ds.csv
# df = pd.read_csv('netflix_ds.csv')
df = pd.read_csv(r"C:\Users\prath\Desktop\Prathamesh\Prathamesh\Learn\DSML\2 Common Core Foundations\Netflix_Project\DataSet\netflix.csv")
df.head()

NameError: name 'pd' is not defined

#2. Observations on the shape of data, data types of all the attributes, conversion of categorical attributes to 'category' (If required), missing value detection, statistical summary (10 Points)

## Observe shape of data, data types of all the attributes

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


---
**Shape:**
- DataSet has `12 columns` (attributes)
- DataSet has `8807 data points`

**Data Type:**
- Attribute `release_year` is of `int64` Dtype
- `All the remaining atributes` are of `object` Dtype. 
---

##Numeric Variable (`release_year`): Missing value detection, statistical summary

In [4]:
df.describe()

,release_year
count,8807.000000
mean,2014.180198
std,8.819312
min,1925.000000
25%,2013.000000
50%,2017.000000
75%,2019.000000
max,2021.000000


In [5]:
df['release_year'].isna().sum()

0

---
**Missing Values:**
- There are no missing Values
- There are no outliers like 0, -999, etc.

**Statistical Summary:**
- Data is from the year `1925` to `2021`
- Data Distribution
  - `50%` of the data is of the shows that are released `from 2017 till 2021` (~5 yrs)
  - `25%` of the data is of the shows that are released `from 2013 till 2017` (~4 yrs)
  - `25%` of the data is of the shows that are released `from 1925 till 2013` (~88 yrs)
---

#Non-Numeric Variables: Data Cleaning, Missing value detection, statistical summary, conversion of categorical attributes to 'category'

**Data Cleaning:**
- **Striping blank spaces** at the start and the end
- Treat blank spaces as missing values

In [6]:
# Striping the values in Non-Numeric columns
df1 = df.drop('release_year', axis=1).apply(lambda x: x.str.rstrip()).apply(lambda x: x.str.lstrip())
# Assign to NaN
df1 = df1.apply(lambda series_: series_.apply(lambda value_: np.nan if value_=='' else value_))
# Update the DF
df[df1.columns] = df1
del df1
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


**Misiing Values**

In [7]:
df.isna().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

---
`No missing values` in columns
- show_id
- type
- title
- listed_in
- description

Some columns have `missing values`
- dirctor:&emsp;&emsp;$\;$2634
- country:&emsp;&emsp;$\;\;$831
- cast:&emsp;&emsp;&emsp;&emsp;825
- date_added:&emsp;$\;$10
- rating:&emsp;&emsp;&emsp;&emsp;$\;$4
- duration:&emsp;&emsp;&emsp;$\;$3
---

**Statistical Summary**

In [8]:
df.describe(include=object)

,show_id,type,title,director,cast,country,date_added,rating,duration,listed_in,description
count,8807,8807,8807,6173,7982,7976,8797,8803,8804,8807,8807
unique,8807,2,8806,4528,7692,748,1714,17,220,514,8775
top,s1,Movie,Consequences,Rajiv Chilaka,David Attenborough,United States,"January 1, 2020",TV-MA,1 Season,"Dramas, International Movies","Paranormal activity at a lush, abandoned prope..."
freq,1,6131,2,19,19,2818,110,3207,1793,362,4


---
**show_id:**
- All the values are unique

**type:**
- It has two categories

**title:**
- There is a duplicate title: 'Consequences'

**director:**
- There are 4528 unique combination of directors
- Rajiv Chilaka alone has directed the most (19) shows

**cast:**
- There are 7692 unique combination of cast
- David Attenborough as a single cast has acted in the most (19) shows

**country:**
- Shows were produced in 748 different combinations of countries
- United States alone has produced the most (2818) shows

**date_added:**
- Out of 8797 available dates, there are 1714 unique date_added
- On January 1, 2020, most (110) shows were added

**rating:**
- There are 17 unique categories of rating
- Most (3207) shows have rating: TV-MA

**duration:**
- OUt of 8797 available durations, there are 220 unique durations
- Season 1 is the most repeated (1793) duration

**listed_in:**
- OUt of 8807 available genre combinations, there are 514 unique genre combinations
- Most of the shows (362) fall under 'Dramas, International Movies' Genre

**description:**
- Most of the shows have unique descriptions
---

**Conversion of Categorical Attributes to 'category'**

- Convert the show type column Dtype to category as it has two categoreis: Movie and TV Show
- Combine the ratings that have same meaning and create new ratings for the analysis. Then convert the rating Dtype to category

In [9]:
# Convert show type to category Dtype
df['type'] = df['type'].astype('category')

# Combine ratings having same meaning
df.loc[df['rating'].isin(['TV-MA', 'R', 'NC-17']), 'rating'] = '17+'
df.loc[df['rating'].isin(['G', 'TV-G']), 'rating'] = 'General'
df.loc[df['rating'].isin(['NR', 'UR']), 'rating'] = 'Not Rated'
df.loc[df['rating']=='TV-Y', 'rating'] = 'All Kids'
df.loc[df['rating'].isin(['TV-PG', 'PG']), 'rating'] = 'Parental Guidance'
df.loc[df['rating']=='PG-13', 'rating'] = '13+'
df.loc[df['rating']=='TV-14', 'rating'] = '14+'
df.loc[df['rating'].isin(['TV-Y7', 'TV-Y7-FV']), 'rating'] = '7+'

# Convert rating column Dtype to category
df['rating'] = df['rating'].astype('category')

#5. Missing Value & Outlier check (Treatment optional) (10 Points)

##For Variable: director, cast and country

To find the best guess for the missing values, I tried to creat diffrent groups and observed for any pattern but I did not find any pattern.

Assign `missing director, cast and country` values as  `Unknown`

In [10]:
df[['director', 'cast', 'country']] = df[['director', 'cast', 'country']].fillna('Unknown')

##For Variable: date_added


Convert the `date_added` column Dtype to `datetime`

To find the best guess for the missing date_added, I tried to creat diffrent groups and observed if ther is any pattern/ correlation with release date, type, directory, cast and country; but I could not find any.

Since missing date_added values are less in numbers (10). I would `keep it as NaN`. 

Check if the `date_added is before the release_year` of any show. If it is so, it is a wrong information. We are not sure if the release_year or date_added is correct. Therefore for such records we make those values `NaN`.

In [11]:
df['date_added'] = pd.to_datetime(df['date_added'])

In [12]:
df.loc[df['date_added'].dt.year < df['release_year'], ['date_added', 'release_year']] = np.nan 

##For Variable: duration

In [13]:
df.loc[df['duration'].isna()]

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
5541,s5542,Movie,Louis C.K. 2017,Louis C.K.,Louis C.K.,United States,2017-04-04,2017.0,74 min,NaN,Movies,"Louis C.K. muses on religion, eternal love, gi..."
5794,s5795,Movie,Louis C.K.: Hilarious,Louis C.K.,Louis C.K.,United States,2016-09-16,2010.0,84 min,NaN,Movies,Emmy-winning comedy writer Louis C.K. brings h...
5813,s5814,Movie,Louis C.K.: Live at the Comedy Store,Louis C.K.,Louis C.K.,United States,2016-08-15,2015.0,66 min,NaN,Movies,The comic puts his trademark hilarious/thought...


---
Duration values of the shows are shifted to rating columns.
Replace the duration from rating column to duration column

---

In [14]:
df.loc[df['duration'].isna(), ['rating', 'duration']] = df.loc[df['duration'].isna(), ['duration', 'rating']].values

Recalculate `rating` categories

In [15]:
df['rating'] = df['rating'].astype('string')
df['rating'] = df['rating'].astype('category')

**Make the duration units uniform**
- Check different units available for type: Movie and TV Show
- Remove the units and convert the Dtype to Numeric

**For type: Movie**

check for different units used

In [16]:
df.loc[df['type']=='Movie', 'duration'].str.split(' ').apply(lambda x: '-'.join([i for i in x if i.isalpha()])).unique()

array(['min'], dtype=object)

---
All the durations of type Movie are in `min`

---

Check if all the values in the durtion column contains only one numeric value

In [17]:
df.loc[df['type']=='Movie', 'duration'].str.split(' ').apply(lambda x: len([i for i in x if i.isnumeric()])==1).all()

True

Keep only the numeric values

In [18]:
df.loc[df['type']=='Movie', 'duration'] = df.loc[df['type']=='Movie', 'duration'].str.split(' ').apply(lambda x: [i for i in x if i.isnumeric()][0])

**For type: TV Show**

check for different units used

In [19]:
df.loc[df['type']=='TV Show', 'duration'].str.split(' ').apply(lambda x: '-'.join([i for i in x if i.isalpha()])).unique()

array(['Seasons', 'Season'], dtype=object)

---
All the durations of type TV Show are in either `Season` for 1 season and `Seasons` for multiple seasons

---

Check if all the values in the durtion column contains only one numeric value

In [20]:
df.loc[df['type']=='TV Show', 'duration'].str.split(' ').apply(lambda x: len([i for i in x if i.isnumeric()])==1).all()

True

Keep only the numeric values

In [21]:
df.loc[df['type']=='TV Show', 'duration'] = df.loc[df['type']=='TV Show', 'duration'].str.split(' ').apply(lambda x: [i for i in x if i.isnumeric()][0])

**Convert the Dtype to integet**

In [22]:
df['duration'] = df['duration'].astype('int')

##For Variable: rating

Assign `Missing Rating` values to `Not Rated` Category

In [23]:
df.loc[df['rating'].isna(), 'rating'] = 'Not Rated'

#Check for missing values

In [24]:
df.isna().sum()

show_id          0
type             0
title            0
director         0
cast             0
country          0
date_added      24
release_year    14
rating           0
duration         0
listed_in        0
description      0
dtype: int64

---
As discussed, all the missing values are treated except the date_added column

---

#Check for outlier values

In [25]:
df.groupby('type').describe()

release_year                                                         \
               count         mean       std     min     25%     50%     75%   
type                                                                          
Movie         6129.0  2013.119759  9.679256  1942.0  2012.0  2016.0  2018.0   
TV Show       2664.0  2016.593468  5.749193  1925.0  2016.0  2018.0  2020.0   

                duration                                                       
            max    count       mean        std  min   25%   50%    75%    max  
type                                                                           
Movie    2021.0   6131.0  99.564998  28.289504  3.0  87.0  98.0  114.0  312.0  
TV Show  2021.0   2676.0   1.764948   1.582752  1.0   1.0   1.0    2.0   17.0

In [26]:
df.groupby('type').describe(include=np.datetime64, datetime_is_numeric=True)

date_added                                           \
             count                          mean        min   
type                                                          
Movie         6129 2019-05-07 06:56:47.929515264 2008-01-01   
TV Show       2654 2019-06-10 13:43:05.380557568 2008-02-04   

                                                                       
                        25%                 50%        75%        max  
type                                                                   
Movie   2018-04-01 00:00:00 2019-06-20 00:00:00 2020-07-24 2021-09-25  
TV Show 2018-04-21 18:00:00 2019-08-15 12:00:00 2020-10-01 2021-09-24

In [27]:
df.groupby('type').describe(include='category')

rating                  
         count unique  top  freq
type                            
Movie     6131      8  17+  2862
TV Show   2676      7  17+  1147

In [28]:
df.groupby('type')[['title', 'director', 'cast', 'country', 'listed_in']].describe(include=object)

title                            director                        cast  \
        count unique            top freq    count unique      top  freq count   
type                                                                            
Movie    6131   6130   Consequences    2     6131   4355  Unknown   188  6131   
TV Show  2676   2676  Blood & Water    1     2676    223  Unknown  2446  2676   

                             country                             listed_in  \
        unique      top freq   count unique            top  freq     count   
type                                                                         
Movie     5446  Unknown  475    6131    652  United States  2058      6131   
TV Show   2284  Unknown  350    2676    197  United States   760      2676   

                                                   
        unique                           top freq  
type                                               
Movie      278  Dramas, International Movies  362  
TV Show    236                      Kids' TV  220

---
Check Movies with title 'Consequences' as there are two movies with same title

---

In [29]:
df.loc[df['title']=='Consequences']

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
3371,s3372,Movie,Consequences,Ozan Açıktan,"Nehir Erdoğan, Tardu Flordun, İlker Kaleli, Se...",Turkey,2019-10-25,2014.0,17+,106,"Dramas, International Movies, Thrillers",Secrets bubble to the surface after a sensual ...
6529,s6530,Movie,Consequences,Ozan Açıktan,"Nehir Erdoğan, Tardu Flordun, İlker Kaleli, Se...",Turkey,2019-10-25,2014.0,17+,106,"Dramas, International Movies, Thrillers",Secrets bubble to the surface after a sensual ...


Remove the duplicate movie record

In [30]:
df = df.drop_duplicates(subset='title')

#3. Non-Graphical Analysis: Value counts and unique attributes ​​(10 Points)

##Types of show (Movie/ TV Show); Unique Counts

In [31]:
df['type'].value_counts()

Movie      6130
TV Show    2676
Name: type, dtype: int64

In [32]:
(df['type'].value_counts()/ df.shape[0] * 100).round(2)

Movie      69.61
TV Show    30.39
Name: type, dtype: float64

In [33]:
df.groupby('type')[['show_id', 'title']].nunique()

,show_id,title
type,,
Movie,6130,6130
TV Show,2676,2676


---
Columns `show_id` and `title` are `unique` for each show

There are `70% Movies` (6130) and `30% TV Shows` (2676)  

---

#Value Counts of Combination of Directors by Show Type

In [34]:
df.loc[df['type']=='Movie', 'director'].value_counts()

Unknown                       188
Rajiv Chilaka                  19
Raúl Campos, Jan Suter         18
Suhas Kadav                    16
Marcus Raboy                   15
                             ... 
Sion Sono                       1
Dennis Rovira van Boekholt      1
Naoto Amazutsumi                1
Jenny Gage                      1
Mozez Singh                     1
Name: director, Length: 4355, dtype: int64

In [35]:
df.loc[df['type']=='TV Show', 'director'].value_counts()

Unknown                              2446
Alastair Fothergill                     3
Hsu Fu-chun                             2
Iginio Straffi                          2
Shin Won-ho                             2
                                     ... 
Juliana Vicente                         1
Chang Chin-jung, Chen Rong-hui          1
Thierry Demaizière, Alban Teurlai       1
Manolo Caro                             1
Michael Cumming                         1
Name: director, Length: 223, dtype: int64

#Value Counts of Conbination of Cast by Show Type

In [36]:
df.loc[df['type']=='Movie', 'cast'].value_counts()

Unknown                                                                                                                                                                               475
Vatsal Dubey, Julie Tejwani, Rupa Bhimani, Jigna Bhardwaj, Rajesh Kava, Mousam, Swapnil                                                                                                13
Samuel West                                                                                                                                                                            10
Jeff Dunham                                                                                                                                                                             7
Kevin Hart                                                                                                                                                                              6
                                                                      

In [37]:
df.loc[df['type']=='TV Show', 'cast'].value_counts()

Unknown                                                                                                                                                             350
David Attenborough                                                                                                                                                   14
Michela Luci, Jamie Watson, Anna Claire Bartlam, Dante Zee, Eric Peterson                                                                                             4
Dave Chappelle                                                                                                                                                        3
You, Reina Triendl, Yoshimi Tokui, Azusa Babazono, Ryota Yamasato, Shono Hayama                                                                                       2
                                                                                                                                                                

#Value Counts of Combination of Country by Show Type

In [38]:
df.loc[df['type']=='Movie', 'country'].value_counts()

United States                            2058
India                                     893
Unknown                                   440
United Kingdom                            206
Canada                                    122
                                         ... 
United Kingdom, Russia, United States       1
Paraguay, Argentina                         1
United Kingdom, Malawi                      1
Austria, Iraq, United States                1
United Arab Emirates, Jordan                1
Name: country, Length: 652, dtype: int64

In [39]:
df.loc[df['type']=='TV Show', 'country'].value_counts()

United States                                    760
Unknown                                          391
United Kingdom                                   213
Japan                                            169
South Korea                                      158
                                                ... 
Belarus                                            1
United Kingdom, Australia                          1
France, Australia, Germany                         1
Australia, New Zealand, United States              1
United States, France, South Korea, Indonesia      1
Name: country, Length: 197, dtype: int64

#Value Counts of Ratings by Show Type

In [40]:
df.loc[df['type']=='Movie', 'rating'].value_counts()

17+                  2861
14+                  1427
Parental Guidance     827
13+                   490
General               167
7+                    144
All Kids              131
Not Rated              83
Name: rating, dtype: int64

In [41]:
df.loc[df['type']=='TV Show', 'rating'].value_counts()

17+                  1147
14+                   733
Parental Guidance     323
7+                    196
All Kids              176
General                94
Not Rated               7
13+                     0
Name: rating, dtype: int64

#Value Counts of Combinations of Gener by Show Type

In [42]:
df.loc[df['type']=='Movie', 'listed_in'].value_counts()

Dramas, International Movies                        362
Documentaries                                       359
Stand-Up Comedy                                     334
Comedies, Dramas, International Movies              274
Dramas, Independent Movies, International Movies    252
                                                   ... 
Sci-Fi & Fantasy                                      1
Sports Movies                                         1
Children & Family Movies, Comedies, Cult Movies       1
Cult Movies, Dramas, Music & Musicals                 1
Cult Movies, Dramas, Thrillers                        1
Name: listed_in, Length: 278, dtype: int64

In [43]:
df.loc[df['type']=='TV Show', 'listed_in'].value_counts()

Kids' TV                                                 220
International TV Shows, TV Dramas                        121
Crime TV Shows, International TV Shows, TV Dramas        110
Kids' TV, TV Comedies                                     99
Reality TV                                                95
                                                        ... 
Kids' TV, TV Action & Adventure, TV Dramas                 1
British TV Shows, Kids' TV, TV Thrillers                   1
Reality TV, TV Horror, TV Thrillers                        1
TV Action & Adventure, TV Horror, TV Sci-Fi & Fantasy      1
Classic & Cult TV, Crime TV Shows, TV Dramas               1
Name: listed_in, Length: 236, dtype: int64

#Distribution of Date Added by Show Type

In [44]:
df.loc[df['type']=='Movie', 'date_added'].describe(datetime_is_numeric=True)

count                             6128
mean     2019-05-07 06:16:41.044386560
min                2008-01-01 00:00:00
25%                2018-04-01 00:00:00
50%                2019-06-19 12:00:00
75%                2020-07-24 00:00:00
max                2021-09-25 00:00:00
Name: date_added, dtype: object

In [45]:
df.loc[df['type']=='TV Show', 'date_added'].describe(datetime_is_numeric=True)

count                             2654
mean     2019-06-10 13:43:05.380557568
min                2008-02-04 00:00:00
25%                2018-04-21 18:00:00
50%                2019-08-15 12:00:00
75%                2020-10-01 00:00:00
max                2021-09-24 00:00:00
Name: date_added, dtype: object

#Distribution of Release Year by Show Type

In [46]:
df.loc[df['type']=='Movie', 'release_year'].describe().round()

count    6128.0
mean     2013.0
std        10.0
min      1942.0
25%      2012.0
50%      2016.0
75%      2018.0
max      2021.0
Name: release_year, dtype: float64

In [47]:
df.loc[df['type']=='TV Show', 'release_year'].describe().round()

count    2664.0
mean     2017.0
std         6.0
min      1925.0
25%      2016.0
50%      2018.0
75%      2020.0
max      2021.0
Name: release_year, dtype: float64

#Distribution of Duration by Show Type

In [48]:
df.loc[df['type']=='Movie', 'duration'].describe().round()

count    6130.0
mean      100.0
std        28.0
min         3.0
25%        87.0
50%        98.0
75%       114.0
max       312.0
Name: duration, dtype: float64

In [49]:
df.loc[df['type']=='TV Show', 'duration'].describe().round()

count    2676.0
mean        2.0
std         2.0
min         1.0
25%         1.0
50%         1.0
75%         2.0
max        17.0
Name: duration, dtype: float64

#4. Visual Analysis - Univariate, Bivariate after pre-processing of the data: Preprocessing

**Define a Function to Split Nested Data Column into Rows**


In [50]:
def unnested_column(df, id_col, nested_col):
  df1 = pd.concat((df[id_col], df[nested_col].str.split(',', expand=True)), axis=1)
  df1 = pd.melt(df1.reset_index(), id_vars=['index', id_col], value_name=nested_col).dropna().sort_values('index').drop(['index', 'variable'], axis=1)
  df = pd.merge(df.drop(nested_col, axis=1), df1, on=id_col)
  return df

**Unnest director, cast, country and listed_in Columns**

In [ ]:
df = unnested_column(df, 'show_id', 'director')
df = unnested_column(df, 'show_id', 'cast')
df = unnested_column(df, 'show_id', 'country')
df = unnested_column(df, 'show_id', 'listed_in')

**Rearrange the Columns**

In [ ]:
df = df[['show_id', 'type', 'title', 'director', 'cast', 'country', 'listed_in', 'duration', 'rating', 'date_added', 'release_year', 'description']]

In [ ]:
df.head(10)

#Testing

In [ ]:
df.head()

In [ ]:
df2 = pd.DataFrame({
    'col1': [' ', ' asdf', 'asd  ', np.nan, '  asdf  ', 'asdf', '', ''],
    'col2': [1, 2, 0, 4, 5, 6, 7, np.nan],
    'col3': ['1', '2', '0', '4', '5', '6', '7', '8']
})
df2

In [ ]:
df2.loc[0,'col1'] is None

In [ ]:
df2.apply(lambda x: x.apply(lambda y: np.nan if y=='' else y))

In [ ]:
def blankToNan(col):
  return col.apply(lambda x: np.nan if x=='' else x)
df2.apply(lambda x: blankToNan(x))

In [ ]:
df2.info()

In [ ]:
df2.loc[:, 'col2'].isna().sum()

In [ ]:
# df2['col1'].str.rstrip().isna()
# df2['col1'].str.rstrip().reset_index(drop=True)
df2.drop('col2', axis=1).apply(lambda x: x.str.rstrip()).info()

In [ ]:
# df.loc[0, 'title'] = '    '+df.loc[0, 'title']+'    '
# df.loc[0, 'title']

In [ ]:
df.loc[0][2][0]